motion/thruster_allocator/src/allocator_ros.cpp

```diff
@@ -0,0 +1,110 @@
+/**
+ * @file allocator_ros.cpp
+ * @brief This file contains the implementation of the Allocator class, which
+ * allocates thrust to the ASV's thrusters based on the desired body frame
+ * forces.
+ */
+
+#include <thruster_allocator/allocator_ros.hpp>
+#include <thruster_allocator/allocator_utils.hpp>
+#include <thruster_allocator/pseudoinverse_allocator.hpp>
+#include <vortex_msgs/msg/thruster_forces.hpp>
+
+#include <chrono>
+#include <functional>
+
+using namespace std::chrono_literals;
+
+ThrusterAllocator::ThrusterAllocator()
+    : Node("thruster_allocator_node"),
+      pseudoinverse_allocator_(Eigen::MatrixXd::Zero(3, 4)) {
+  subscription_ = this->create_subscription<geometry_msgs::msg::Wrench>(
+      "thrust/wrench_input", 1,
+      std::bind(&ThrusterAllocator::wrench_callback, this,
+                std::placeholders::_1));
+
+  publisher_ = this->create_publisher<vortex_msgs::msg::ThrusterForces>(
+      "thrust/thruster_forces", 1);
+
+  timer_ = this->create_wall_timer(
+      100ms, std::bind(&ThrusterAllocator::timer_callback, this));
+
+  Eigen::MatrixXd thrust_configuration_pseudoinverse;
+  calculateRightPseudoinverse(thrust_configuration,
+                              thrust_configuration_pseudoinverse);
+
+  // Set T_pinv in pseudoinverse_allocator_
+  pseudoinverse_allocator_.T_pinv = thrust_configuration_pseudoinverse;
+
+  body_frame_forces_.setZero();
+}
+
+/**
+ * @brief This function is called by the timer and calculates the allocated
+ * thrust based on the body frame forces. It then saturates the output vector
+ * between min and max values and publishes the thruster forces to the topic
+ * "thrust/thruster_forces". If the calculated thruster forces vector is
+ * invalid, it logs an error message. If the thruster forces vector required
+ * saturation, it logs a warning message.
+ */
```

Comment:
This is not very brief. It should only be a short description of the core functionality. Logging of error messages is not really relevant here.